### PR TITLE
New page: color modules example

### PR DESCRIPTION
**Pull request by estelle**
Example for new CSS color module page.

### CODE REVIEW

File: modules/colors.html [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    h = (( g - b ) / delta ) % 6;
+  else if ( cmax == g )
+    h = ( b - r ) / delta + 2;
+  else
+    h = ( r - g ) / delta + 4;
+
+  h = Math.round( h * 60 );
+
+  if ( h < 0 )
+    h += 360;
```
Comment (teoli2003) [inline modules/colors.html:133]:
```suggestion
  if ( delta === 0 ) {
    h = 0;
  } else if ( cmax === r ) {
    h = (( g - b ) / delta ) % 6;
  } else if ( cmax === g ) {
    h = ( b - r ) / delta + 2;
  } else {
    h = ( r - g ) / delta + 4;
  }
  
  h = Math.round( h * 60 );

  if ( h < 0 ) {
    h += 360;
  }
```

File: modules/colors.html [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  const cell = document.querySelector( "#HWB td" );
+  const hsv = s * ( l < 50 ? l : 100 - l ) / 100;
+	const W = hsv === 0 ? 0 : 2 * hsv / ( l + hsv ) * 100;
+	const B = l + hsv;
+  if( opacity == 1 ) {
+    cell.innerHTML = `hwb(${h} ${W.toFixed( 1 )}% ${B.toFixed( 1 )}%)`;
+  }
+  else {
+    cell.innerHTML = `hwb(${h} ${W.toFixed( 1 )}% ${B.toFixed( 1 )}% / ${opacity})`;
+  }
```
Comment (teoli2003) [inline modules/colors.html:158]:
```suggestion
  if( opacity === 1 ) {
    cell.textContent = `hwb(${h} ${W.toFixed( 1 )}% ${B.toFixed( 1 )}%)`;
  }
  else {
    cell.textContent = `hwb(${h} ${W.toFixed( 1 )}% ${B.toFixed( 1 )}% / ${opacity})`;
  }
```

File: modules/colors.html [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+const setBackgroundColor = function ( color, opacity ) {
```
Comment (teoli2003) [inline modules/colors.html:161]:
```suggestion
function setBackgroundColor( color, opacity ) {
```

File: modules/colors.html [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+  const body = document.querySelector( "div" );
+  if( opacity == 1 ) {
```
Comment (teoli2003) [inline modules/colors.html:163]:
```suggestion
  if( opacity === 1 ) {
```

File: modules/colors.html [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    body.style.backgroundColor = color;
+  } else {
+    body.style.backgroundColor = color + "" + hexOpacity( opacity );
```
Comment (teoli2003) [inline modules/colors.html:166]:
```suggestion
    body.style.backgroundColor = `${color}${hexOpacity(opacity)}`;
```

File: modules/colors.html [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+  }
+}
+
+const hexOpacity = function( opacity ) {
```
Comment (teoli2003) [inline modules/colors.html:170]:
```suggestion
function hexOpacity( opacity ) {
```

**Review by teoli2003 — CHANGES_REQUESTED** [review]
Made a first pass. We also need to run Prettify on this file; I'll do a last pass afterward (it is difficult to read like this).

Unless you know why, you should always use `===` instead of `==` and `textContent` instead of `innerHTML` if there is no HTML involve.

File: modules/colors.html [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+          cell.textContent = `hwb(${h} ${W.toFixed(1)}% ${B.toFixed(
+            1
+          )}% / ${opacity})`;
+        }
+      }
+
+      function setBackgroundColor(color, opacity) {
+        const body = document.querySelector("div");
+        if (opacity !== 1) {
+          color = color + "" + hexOpacity(opacity);
```
Comment (estelle) [inline modules/colors.html:190]:
```suggestion
          color = `${color}${hexOpacity(opacity)}`;
```

**Comment by estelle** [comment]
@teoli2003 We went with 

```javascript
function hexOpacity(color, opacity)  {
  let char = "00";
  if (opacity > 0) {
    char = Math.floor(opacity * 255).toString(16);
  }
  return `${color}${char}`;
}
```

Is it also ok to go with the following way of declaring a function (the first line), or against our best practices guidelines?

```javascript
const hexOpacity = (color, opacity) => {
  let char = "00";
  if (opacity > 0) {
    char = Math.floor(opacity * 255).toString(16);
  }
  return `${color}${char}`;
}
```

File: modules/colors.html [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+        createColor();
+      });
+
+      opacityPicker.addEventListener("change", () => {
+        createColor();
+      });
+
+      function createColor() {
+        const currentColor = colorPicker.value;
+        const currentOpacity = opacityPicker.value * 1;
```
Comment (teoli2003) [inline modules/colors.html:116]: See https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Writing_style_guide/Code_style_guide/JavaScript#type_coercion
```suggestion
        const currentOpacity = Number(opacityPicker.value);
```

File: modules/colors.html [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+          cell.textContent = `rgb(${R} ${G} ${B})`;
+        } else {
+          cell.textContent = `rgb(${R} ${G} ${B} / ${opacity})`;
+        }
+        createHSL(R, G, B, opacity);
+      }
+
+      function createHSL(r, g, b, opacity) {
+        const cell = document.querySelector("#HSL td");
+        (r = r / 255), (g = g / 255), (b = b / 255);
```
Comment (teoli2003) [inline modules/colors.html:149]: This is difficult to read.
```suggestion
        // Let's have r, g, b in the range [0, 1]
        r = r / 255;
        g = g / 255;
        b = b / 255;
```

File: modules/colors.html [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+        const cmax = Math.max(r, g, b);
+        const delta = cmax - cmin;
+        let h = 0,
+          s = 0,
+          l = 0;
+
+        if (delta === 0) h = 0;
+        else if (cmax === r) h = ((g - b) / delta) % 6;
+        else if (cmax === g) h = (b - r) / delta + 2;
+        else h = (r - g) / delta + 4;
```
Comment (teoli2003) [inline modules/colors.html:160]: According our guidelines:
```suggestion
        if (delta === 0) {
          h = 0;
        } else if (cmax === r) {
          h = ((g - b) / delta) % 6;
        } else if (cmax === g) {
          h = (b - r) / delta + 2;
        }
        else h = (r - g) / delta + 4;
```

File: modules/colors.html [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+
+        h = Math.round(h * 60);
+
+        if (h < 0) h += 360;
```
Comment (teoli2003) [inline modules/colors.html:164]:
```suggestion
        // We want an angle between 0 and 360°
        if (h < 0) {
          h += 360;
        }
```

File: modules/colors.html [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+
+        l = (cmax + cmin) / 2;
+        s = (delta === 0) ? 0 : delta / (1 - Math.abs(2 * l - 1));
+        s = +(s * 100).toFixed(1);
+        l = +(l * 100).toFixed(1);
```
Comment (teoli2003) [inline modules/colors.html:169]: If there is a risk that the result of toFixed() is not a number, we should convert it explicitely:
```suggestion
        s = Number((s * 100).toFixed(1));
        l = Number((l * 100).toFixed(1));
```

File: modules/colors.html [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+        const hsv = (s * ((l < 50) ? l : 100 - l)) / 100;
+        const W = (hsv === 0) ? 0 : ((2 * hsv) / (l + hsv)) * 100;
+        const B = l + hsv;
+        if (opacity === 1) {
+          cell.textContent = `hwb(${h} ${W.toFixed(1)}% ${B.toFixed(1)}%)`;
+        } else {
+          cell.textContent = `hwb(${h} ${W.toFixed(1)}% ${B.toFixed(
+            1
+          )}% / ${opacity})`;
+        }
```
Comment (teoli2003) [inline modules/colors.html:190]: I think this is easier to read:
```suggestion
        let W = (hsv === 0) ? 0 : ((2 * hsv) / (l + hsv)) * 100;
        let B = l + hsv;
        W = W.toFixed(1);
        B = B.toFixed(1);
        if (opacity === 1) {
          cell.textContent = `hwb(${h} ${W}% ${B}%)`;
        } else {
          cell.textContent = `hwb(${h} ${W}% ${B}% / ${opacity})`;
        }
```

**Review by teoli2003 — CHANGES_REQUESTED** [review]
There are a few last small changes, but I find this much easier to read already!

**Comment by teoli2003** [comment]
> @teoli2003 We went with
> 
> ```js
> function hexOpacity(color, opacity)  {
>   let char = "00";
>   if (opacity > 0) {
>     char = Math.floor(opacity * 255).toString(16);
>   }
>   return `${color}${char}`;
> }
> ```
> 
> Is it also ok to go with the following way of declaring a function (the first line), or against our best practices guidelines?
> 
> ```js
> const hexOpacity = (color, opacity) => {
>   let char = "00";
>   if (opacity > 0) {
>     char = Math.floor(opacity * 255).toString(16);
>   }
>   return `${color}${char}`;
> }
> ```

When possible, I prefer to have a `function` definition. I don't see what the longest syntax brings to readability.

(I'm not a fan of setting the else value before the if, but that's not against our rules and a matter of taste)

**Review by teoli2003 — APPROVED** [review]
👍

**Pull request merged**